### PR TITLE
updated index calculation

### LINKED ISSUE #166: Index depends on length
In the second crypto review (#164) there was a note that plaintexts of different lengths might end up with similar results, if the extra bits were zeros. The suggestion to fix this was to update the path index to also include the length.

### PR DESCRIPTION
**Pull request by martinduke**
Fix #166 in accordance with the 2nd crypto review

### CODE REVIEW

File: draft-ietf-quic-load-balancers.md [hunk 1]
```diff
@@ -450,45 +450,47 @@ length, splitting an odd octet in half if necessary. For example,
...
 
 4. XOR the least significant bits of the ciphertext with right_0 to form
 right_1.
 
     Thus steps 3 and 4 can be expressed as
     ```
     right_1 = right_0 ^ truncate_right(
-                            AES_ECB(key, expand_left(left_0, 0x01)),
+                            AES_ECB(key, expand_left(left_0, 1 + cid_len)),
                             len(right_0))
```
Comment (huitema) [inline draft-ietf-quic-load-balancers.md:464]: "1+cid_len" is not ideal, especially if we have to bump the count to 12. I would rather have something "len on 8 bits + rank on 4 bits"

**Review by huitema — COMMENTED** [review]
I don't like replacing "0x01", which specifies bits, by "1 + cid_len". Several issues:
* is the encoding of the integer big indian or little endian?
* how many bits?
* could cid_len1 + pass_number_x collide with cid_len2 + pass_number_y
I would feel more comfortable with "pass_number + 16*cid_len" + specify the bits.

File: draft-ietf-quic-load-balancers.md (hunk 1 above)
Comment (martinduke) [reply draft-ietf-quic-load-balancers.md:464]: I was envisioning all this fitting in one byte (which was not clear in the text), but I agree parsing the space is a better result.

File: draft-ietf-quic-load-balancers.md [hunk 2]
```diff
@@ -450,45 +450,48 @@ length, splitting an odd octet in half if necessary. For example,
...
-results with left_1.
+encrypting right_2 with the most significant octet as the concatenation of
+0b11 ands cid_len, and XOR the results with left_1.
 
     ```
     left_2 = left_1 ^ truncate_left(
-                          AES_ECB(key, expand_right(right_2, 0x04)),
+                          AES_ECB(key, expand_right(right_2, 0xc0 | cid_len)),
                           len(left_1))
     ```
```
Comment (huitema) [inline draft-ietf-quic-load-balancers.md:496]: So the index is effectively:

* 3 bits for the identifier of the step,
* 5 bits for the length of the CID.

If we stick with QUIC v1, the max CID length is 20 bytes. 5 bits definitely suffice. But if we ever increase the number of steps to 12, 3 bits will not suffice.

Do we really need to pack bits? The "right" and "left" fields are at most 10 bytes, so we have 6 bytes available for the "tweak". Can't we have 1 byte for the step number and one byte for the CID length? 
 The "right" and "left" can be at most 10 bytes.

**Review by huitema — COMMENTED** [review]
That works. My main comment below is that you are packing bits, leaving only 3 to encode the step number. That will work for up to 8 steps, but we have a worst case of 12 steps. Do we need to pack bits, or should we just use 2 bytes?

File: draft-ietf-quic-load-balancers.md [hunk 3]
```diff
@@ -416,19 +416,20 @@ functions that minimize the amount of bit-shifting necessary in the event that
...
 expand_right() is similar, except that the second argument is in the most
-significant byte, and the first argument is in the least significant bits.
-Therefore,
+significant byte, the third is in the second most significant byte, and the
+first argument is in the least significant bits. Therefore,
 
 ~~~pseudocode
-expand_right(0xaaba3c, 0x13) = 0x13000000000000000000000000aaba3c
+expand_right(0xaaba3c, 0x0b, 0x02 = 0x0b020000000000000000000000aaba3c
 ~~~
```
Comment (huitema) [inline draft-ietf-quic-load-balancers.md:434]: Missing parenthesis. Should be `expand_right(0xaaba3c, 0x0b, 0x02)`.

File: draft-ietf-quic-load-balancers.md (hunk 2 above)
Comment (huitema) [reply draft-ietf-quic-load-balancers.md:497]: This comment is addressed in the new text.

**Review by huitema — APPROVED** [review]
Looks good, except for the missing parenthesis.

**Pull request merged**